src/generic_event_type.cpp

```diff
@@ -0,0 +1,374 @@
+    }
+    for( efftype_id effect : effects_to_remove ) {
+        if( target.has_effect( effect ) ) {
+            target.remove_effect( effect );
+        }
+    }
+    for( effect_info effect : effects_to_add ) {
+        target.add_effect( effect.id, effect.length, effect.target_part->token, effect.intensity );
+    }
+    for( bionic_id cbm : cbms_to_add ) {
+        if( !target.has_bionic( cbm ) ) {
+            target.add_bionic( cbm );
+        }
+    }
+    for( bionic_id cbm : cbms_to_remove ) {
+        if( target.has_bionic( cbm ) ) {
+            target.remove_bionic( cbm );
+        }
+    }
+    for( std::string variable : generic_variables_to_set_true ) {
+        g->generic_variable_map[variable] = true;
+    }
+    for( std::string variable : generic_variables_to_set_false ) {
+        g->generic_variable_map[variable] = false;
+    }
+    for( std::pair<time_duration, generic_event_type_id> pair : events_to_queue ) {
+        generic_event_types::queue_generic_event( pair.first, pair.second );
+    }
+    if( damage != 0 ) {
+        if( target_part.is_valid() ) {
+            target.deal_damage( nullptr, target_part, damage_instance( DT_BASH,
+                                damage ) );
+        } else {
+            for( const bodypart_id &bp : target.get_all_body_parts() ) {
+                target.deal_damage( nullptr, bp, damage_instance( DT_BASH, damage ) );
+            }
+        }
+    }
+    if( weather_change != WEATHER_NULL ) {
+        get_weather().weather_override = weather_change;
+    }
+    if( update_weather ) {
+        get_weather().next_weather = true;
+    }
+
+    target.mod_healthy( healthy );
+    target.mod_rad( radiation );
+    wet_character( target, wet );
+    target.mod_pain( pain );
+    sound( sound_message, sound_effect );
+    target.add_msg_if_player( message );
+}
+
+void generic_event_types::load_pair( const JsonObject &jo, const std::string & )
+{
+    g->generic_events_vector.emplace_back( generic_requirement_type_id( jo.get_string( "require" ) ),
+                                           generic_event_type_id( jo.get_string( "event" ) ) );
+}
+
+void generic_event_types::process_generic_pairs()
+{
+    std::vector<std::pair<time_point, generic_event_type_id>>::iterator queued_event =
+                g->queued_generic_events.begin();
+
+    while( queued_event != g->queued_generic_events.end() ) {
+        if( queued_event->first <= calendar::turn ) {
+            queued_event->second->do_event( );
+            queued_event = g->queued_generic_events.erase( queued_event );
+        } else {
+            ++queued_event;
+        }
+    }
+    for( std::pair<generic_requirement_type_id, generic_event_type_id> require_event :
+         g->generic_events_vector ) {
+        if( require_event.first->test( get_player_character().pos(), get_player_character(),
+                                       WEATHER_NULL ) ) {
+            require_event.second->do_event( );
+        }
+    }
```

Comment:
This has the potential to become quite expensive when run every turn.  We don't need a more performant implementation in this initial PR, but I'd like to know if you've thought about how to make it more performant if necessary.  If design changes are needed then we should make them before we merge this.

Reply:
So  I have thought some about performance.  My biggest thought is to simply default the once_every generic_requirement value to be 5 minutes so all generic_events are only checked once every 5 minutes unless otherwise specified.  It's the first thing to be checked to ensure minimal cost to using it.  Beyond that I'd figured possibly some sort of cache for heavily used generic_requirements but that only starts to make sense when this system is being used to know the performance pain points.

Reply:
That sounds reasonable.  I suggest furthermore that we allow for the possibility that each requirement have a random offset within the five minute period, so that they are spread out rather than all checked together once every five minutes.  No need to implement that now, but in your documentation it might be worth a comment that the exact timing of the "every 5 minutes" is flexible.

If almost all requirements have an "every x amount of time" component, then a reasonable solution would be to handle all those via the queue, rather than checking every time.  Again, not something that has to happen now, but a good opportunity for improved performance later.

Reply:
Yes, that seems like a reasonable approach.  In short, whenever a condition fails, shelve that precondition until something changes which could make it true.  For time-based conditions that means the queue; for status-based conditions it might mean the event bus.

Of course, that gets trickier if we support "x OR y" style conditions, which we probably should.